### PR TITLE
Add syntax sugar for class and style attributes

### PR DESCRIPTION
**Pull request by zerg000000**
add code to support attribute value is map or vector.

Case: vector

```{clojure}
[:span.foo {:class ["bar" "baz"]} "text"]
<span class="foo bar baz">text</span>
```

Case: map

```{clojure}
[:span.foo {:style {:background-color "red" :color "red"}} "text"]
<span class="foo" style="background-color:red;color:red;">text</span>
```

### CODE REVIEW

File: src/hiccup/compiler.clj [hunk 1]
```diff
@@ -13,17 +14,33 @@
...
+    (util/escape-html (util/as-str value))
+    (map? value)
+    (str
+      (->> value
+           (mapv (fn [[k v]] (str (util/as-str k) ":" (util/as-str v))))
+           sort
+           (string/join ";"))
+      ";")
+    (sequential? value)
+    (string/join " " value)
```
Comment (weavejester) [inline src/hiccup/compiler.clj]: No escape-html here. The tests didn't fail so we need to test for this as well.

File: src/hiccup/compiler.clj [hunk 2]
```diff
@@ -13,17 +14,33 @@
...
+    :default
```
Comment (weavejester) [inline src/hiccup/compiler.clj]: Typically we use `:else` instead of `:default`.

File: src/hiccup/compiler.clj [hunk 3]
```diff
@@ -13,17 +14,33 @@
 (defn- end-tag []
   (if (xml-mode?) " />" ">"))
 
+(defn- render-attr-value [value]
+  (cond
+    (string? value)
```
Comment (weavejester) [inline src/hiccup/compiler.clj]: This clause is covered by the default case, no?

File: src/hiccup/compiler.clj [hunk 4]
```diff
@@ -13,17 +14,33 @@
...
+    (util/escape-html (util/as-str value))))
+
 (defn- xml-attribute [name value]
-  (str " " (util/as-str name) "=\"" (util/escape-html value) "\""))
+  (str " " (util/as-str name) "=\"" (render-attr-value value) "\""))
 
 (defn- render-attribute [[name value]]
   (cond
     (true? value)
-      (if (xml-mode?)
```
Comment (weavejester) [inline src/hiccup/compiler.clj]: Formatting changes must be removed from the PR. The PR should cover only the changes necessary to implement the feature.

File: src/hiccup/compiler.clj [hunk 5]
```diff
@@ -49,10 +66,17 @@
   (or content
       (and (html-mode?) (not (void-tags tag)))))
 
+(defn class-list [value]
```
Comment (weavejester) [inline src/hiccup/compiler.clj]: This function doesn't seem relevant to the PR either? Could you explain its purpose?

File: src/hiccup/compiler.clj [hunk 6]
```diff
@@ -172,13 +196,13 @@
   [[tag attrs & content :as element]]
   (cond
     (every? literal? element)
-      ::all-literal                    ; e.g. [:span "foo"]
```
Comment (weavejester) [inline src/hiccup/compiler.clj]: More formatting changes that need to be removed, I'm afraid.

**Comment by weavejester** [comment]
Thanks for the commits. I've added some comments in line. Might be a while before I get around to merging this because I haven't focused on Hiccup for a while, but it might be time for me to get back into working on it.

File: src/hiccup/compiler.clj (hunk 6 above)
Comment (zerg000000) [inline src/hiccup/compiler.clj]: sorry about the formatting. Parinfer mode auto formatted the code.

File: src/hiccup/compiler.clj (hunk 1 above)
Comment (zerg000000) [inline src/hiccup/compiler.clj]: The `map?` clause should also be html escaped.

File: src/hiccup/compiler.clj (hunk 5 above)
Comment (weavejester) [inline src/hiccup/compiler.clj]: That makes sense, but I think the way you go about it could be a little more performant. Perhaps something like:

```clojure
(defn- merge-classes [class classes]
  (cond
    (nil? class)    classes
    (string? class) (str class " " classes)
    :else           (str (str/join " " class) " " classes)

(defn- merge-attributes [map-attrs id classes]
  (-> map-attrs
      (cond-> id      (assoc :id id))
      (cond-> classes (assoc :class (merge-classes (:class map-attrs) classes)))))

 (defn normalize-element
   "Ensure an element vector is of the form [tag-name attrs content]."
   [[tag & content]]
   (when-not (or (keyword? tag) (symbol? tag) (string? tag))
      (throw (IllegalArgumentException. (str tag " is not a valid element name."))))
    (let [[_ tag id class] (re-matches re-tag (util/as-str tag))
          classes          (if class (str/replace class "." " "))
          map-attrs        (first content)]
      (if (map? map-attrs)
        [tag (merge-attributes map-attrs id classes) (next content)]
        [tag {:id id, :class classes} content])))

File: src/hiccup/compiler.clj (hunk 1 above)
Comment (weavejester) [inline src/hiccup/compiler.clj]: Yes, though you could just factor the `escape-html` out entirely, since every condition will require it.

**Comment by zerg000000** [comment]
Hi @weavejester , 

any further changes are needed to merge to pull request?

**Comment by weavejester** [comment]
Yep, you still need to address one of my comments: https://github.com/weavejester/hiccup/pull/139#discussion_r92187680

Also I usually abbreviate `clojure.string` to `str`. It's the convention I use in the rest of the Hiccup source, so it would be good to adopt it in the compiler namespace as well.

The code:

```clojure
      (str
        (->> value
             (mapv (fn [[k v]] (str (util/as-str k) ":" v)))
             sort
             (string/join ";"))
        ";")
```

Could be spun off into a separate function for clarity, and we could write it a little better as:

```clojure
(->> value
     (map (fn [[k v]] (str (util/as-str k) ":" v ";")))
     (sort)
     (apply str))

File: src/hiccup/compiler.clj (hunk 5 above)
Comment (zerg000000) [inline src/hiccup/compiler.clj]: cool, much better than my solution. will apply to the pull request

File: src/hiccup/compiler.clj [hunk 7]
```diff
@@ -13,8 +14,20 @@
 (defn- end-tag []
   (if (xml-mode?) " />" ">"))
 
+(defn- render-attr-value [value]
+  (cond
+    (map? value)
+      (->> value
+           (mapv (fn [[k v]] (str (util/as-str k) ":" v ";")))
```
Comment (weavejester) [inline src/hiccup/compiler.clj]: Why `mapv` and not `map`?

File: src/hiccup/compiler.clj [hunk 8]
```diff
@@ -13,8 +14,20 @@
 (defn- end-tag []
   (if (xml-mode?) " />" ">"))
 
+(defn- render-attr-value [value]
+  (cond
+    (map? value)
+      (->> value
```
Comment (weavejester) [inline src/hiccup/compiler.clj]: Can you factor the "map" clause out into its own function for readability?

Comment (zerg000000) [inline src/hiccup/compiler.clj]: yes. may be name it `render-style-map` as it is only used for style currently.

**Comment by zerg000000** [comment]
hi @weavejester , could you help to review this changes? should be completed.

**Comment by weavejester** [comment]
At a glance it looks fine, but I won't be able to properly review and merge it until January.

**Comment by zerg000000** [comment]
noted. ;-)

**Comment by iku000888** [comment]
This has been a much wanted feature, so thanks for the work!

**Comment by weavejester** [comment]
This looks fine. Can you squash down the commits? Perhaps with a commit message like:

    Add syntax sugar for class and style attributes

**Comment by zerg000000** [comment]
Yes. Please check.

p.s. first time to do squash commits.

**Comment by weavejester** [comment]
Looks fine. Merged. :)

**Comment by pesterhazy** [comment]
Wonderful, thanks @zerg000000 and @weavejester for making this happen!

**Comment by theronic** [comment]
Which version of Hiccup supports this? I've been using "1.0.5" as stated on the README, but that does not seem to support it.

**Comment by weavejester** [comment]
@theronic Try 2.0.0-alpha1

**Comment by metasoarous** [comment]
Hello @weavejester. First of all, thanks for your work on this.

Unfortunately, I tried switching to both `2.0.0-alpha1` and `2.0.0-aplha2` (confirming versions with `lein deps :tree`), but wasn't able to get this to work. Am I missing something?

Thanks

**Comment by weavejester** [comment]
Could you give a little more information, @metasoarous? Can you replicate the following results:

```clojure
user=> (require '[hiccup.core :as h])
nil
user=> (h/html [:span.foo {:style {:background-color "red" :color "red"}} "text"]))
"<span class=\"foo\" style=\"background-color:red;color:red;\">text</span>"
```

**Comment by metasoarous** [comment]
@weavejester I _think_ the reason this didn't work for me is that I was in a checkout project, and as a result the environment wasn't getting built properly. At least that's my best guess at the moment. I had some other dependencies misbehaving which suddenly now are fine, after moving the project out of the checkouts directory and re-cleaning things. It's either that or ghosts. In any case, I have now verified that the intended behavior does indeed now work, so I'm good to go. Thanks for hearing me out and taking the time to respond.
